Addon-Controls: Next-generation knobs (#10834)
storybookjs/storybook: addons/controls/README.md

Most importantly, Controls incorporates lessons from years of supporting Knobs on tens of thousands of projects and dozens of different frameworks. We couldn't incrementally fix knobs, so we built a better version.

- [Get started](#get-started)
- [Installation](#installation)
- [Writing stories](#writing-stories)
  - [Auto-generated args](#auto-generated-args)
  - [Manually-configured args](#manually-configured-args)
- [Migrating from knobs](#migrating-from-knobs)
  - [Knobs to auto-generated args](#knobs-to-auto-generated-args)
  - [Knobs to manually-configured args](#knobs-to-manually-configured-args)
- [Available controls](#available-controls)
- [Parameters](#parameters)
    - [expanded](#expanded)
- [FAQs](#faqs)
  - [How will this replace addon-knobs?](#how-will-this-replace-addon-knobs)

## Get started

To get started with `addon-controls`:

1. [install the addon](#installation)
2. [write an args story](#writing-stories)
3. [configure the addon (optional)](#configuration)

If you're installing it into an existing project that uses `addon-knobs`, you might also [migrate your existing stories](#migrating-from-knobs).

## Installation

> In the first version, Controls requires that you're [Storybook Docs](https://github.com/storybookjs/storybook/tree/master/addons/docs). If you're not using it already, please install that first.

Reviewer: ```suggestion
In the first version, Controls requires that you're using [Storybook Docs](https://github.com/storybookjs/storybook/tree/master/addons/docs). If you're not using it already, please install that first.
```

Next, install the package:

```sh
npm install @storybook/addon-controls -D # or yarn
```

And add it to your `.storybook/main.js` config:

```js
module.exports = {
  addons: ['@storybook/addon-controls'],
};
```

Once the addon is installed you should see a `Controls` tab in the addons panel:

## Writing stories

Controls are only available for stories that make use of [Storybook Args](https://github.com/storybookjs/storybook/blob/next/docs/src/pages/formats/component-story-format/index.md#args-story-inputs), so if your story functions don't accept an `Args` object as the first argument, you'll need to upgrade your stories.

Here's a `react` story that doesn't use args:

```js
export const Basic = () => <Button label="hello" />;
```

And here's one that does:

```js

> export const Basic = (args = { label: 'hello' }) => <Button {...args} />;

Reviewer: Does this work? I thought we didn't support this (default value) yet.

Reviewer 2: Normally the storyFn would be called with the storyContext right? So the storyFn has access to parameters etc? Is this now the second argument sometimes?

Author: @ndelangen it's the second argument by default, unless you set the parameter `passArgsFirst: false`

```

There are a few different ways to get from here to there:

- The quickest way to do this is to use [auto-generated args](#auto-generated-args).
- If you're implementing custom story logic based on control value, you should use [manually-configured args](#manually-configured-args)
- And if you're adding this to an existing project that's already using knobs, you should [migrate it](#migrating-from-knobs).

### Auto-generated args

> In the example above, [Storybook Docs](https://github.com/storybookjs/storybook/tree/next/addons/docs) automatically annotates the `Button` with `ArgTypes` metadata that allows `Controls` to automatically create rows for that component.

Author: @yannbf none of the relative paths work on npm 😭

So if the full story was:

```js
import { Button } from './Button';
export default { title: 'Button', component: Button };

export const Basic = () => <Button label="hello" />;
```

> You could simply rewrite it as:

Reviewer 3: This should be changed, according to point 2 of [these guidelines](https://github.com/storybookjs/storybook/tree/next/docs#guidelines-for-writing-good-documentation) of Storybook

```js

> export const Basic = (args = { label: 'hello' }) => <Button {...args} />;

Reviewer 3: Don't forget to add `import { Button } from './Button';` to this example too

Author: It just makes the page much longer without providing informational value. WDYT?

Reviewer 3: I'm not sure, the docs talk about automatic stuff and you're showing scenario A to B, where in scenario A you have an import and in scenario B you don't.. so it might be confusing for people without context. I'd say either:
1 - Remove the import from both, people can infer that the Button is coming from somewhere
2 - Make the text more specific like: `Rewrite the Basic story as:`

```

### Manually-configured args

Auto-generated args make sense when the controls correspond exactly to component. But there are cases where you want controls to control something about the story instead of (or in addition to) the component properties.

Consider the following example that shows reflow behavior of an array of `Button` instances inside a `ButtonGroup` container:

```js
import range from 'lodash/range';
import { Button, ButtonGroup } from '.';

export const Reflow = ({ count }) => (
  <ButtonGroup>
    {range(count).map((i) => (
      <Button label={`button ${i}`} />
    ))}
  </ButtonGroup>
);
```

> In this case, we need to specify `ArgTypes` metadata about the story to render properly:

Reviewer 3: According to the (guidelines)[https://github.com/storybookjs/storybook/tree/next/docs#guidelines-for-writing-good-documentation] of Storybook, avoid the use of pronouns

Author: I disagree with that guideline

Reviewer 3: Then I guess.. let's discuss about them and make them in a way that matches the preferred style so that anyone following them will be consistent with the rest of the documentation?

```js
Reflow.story = {
  argTypes: {
    count: { control: { type: 'range', min: 0, max: 100 } },
  },
};
```

> This is the same kind of data that gets filled in automatically in the [auto-generated args case](#auto-generated-args).

Reviewer 3: > This is the same kind of data that gets filled in automatically in the [auto-generated args case](#auto-generated-args).

This part made me confused. In the text a few lines above it's mentioned that manually configured args are used for special cases where auto generated args don't apply, but this line says that it's the same kind of data that is generated by auto generated args?

## Migrating from knobs

If you're already using [Storybook Knobs](https://github.com/storybookjs/storybook/tree/master/addons/knobs) you should consider migrating to Controls.

> 're probably using it for something that can be satisfied by one of the two previous cases, or [manually-configured args](#manually-configured-args).

Reviewer: ```suggestion
You're probably using it for something that can be satisfied by one of the two previous cases, or [manually-configured args](#manually-configured-args).
```

```js
import { text } from '@storybook/addon-knobs';
import { Button } from './Button';

export const Basic = () => <Button label={text('Label', 'hello')} />;
```

This fills in the Button's label based on a knob, which is exactly the [auto-generated](#auto-generated-args) use case above. So we can rewrite it using auto-generated args:

```js
export const Basic = (args = { label: 'hello' }) => <Button {...args} />;
```

### Knobs to manually-configured args

Similarly, we can also consider a story that uses knob inputs to change its behavior:

```js
import range from 'lodash/range';
import { number } from '@storybook/addon-knobs';

export const Reflow = () => {
  const count = number('Count', 10, { min: 0, max: 100, range: true });
  return (
    <ButtonGroup>
      {range(count).map((i) => (
        <Button label={`button ${i}`} />
      ))}
    </ButtonGroup>
  );
};
```

And again, as above, this can be rewritten using [manually-configured args](#manually-configured-args):

```js
export const Reflow = ({ count }) => (
  <ButtonGroup>
    {range(count).map((i) => (
      <Button label={`button ${i}`} />
    ))}
  </ButtonGroup>
);
Reflow.story = {
  argTypes: {

>     count: { control: { type: 'range', min: 0, max: 100 } },

Reviewer 3: I think there is a mixed message in this example, because not only you're adding the argTypes, but also there seems to be a new mechanic within the control API that translates perfectly the use case of `lodash/range`. Users who don't fit in this use case might either be confused about this change or get it and wonder what other types are available

  },
};
```

## Available controls

> FIXME

Reviewer 3: Oops 👽

## Parameters

> Controls supports the following configuration parameters, either [globally or on a per-story basis](https://storybook.js.org/docs/basics/writing-stories/#parameters):

Reviewer 3: Out of scope of this PR, but the link goes to a page that uses the old parameters way (addParameters) and in line 213 it's using the CSF style, leading to a bit of inconsistency

#### expanded

Since Controls is built on the same engine as Storybook Docs, it can also show property documentation alongside your controls using the `expanded` parameter (defaults to `false`).

To enable expanded mode globally, add the following to `.storybook/preview.js`:

```js
export const parameters = {
  controls: { expanded: true },
};
```

And here's what the resulting UI looks like:

> > FIXME

Reviewer 3: Oops 👽

<center>
  <img src="./docs/media/addon-controls-hero.gif" width="100%" />
</center>

<h1>Storybook Controls</h1>

> Storybook Controls is UI to interactively edit your component stories. Stories written for controls are:

Reviewer 4: ```suggestion
Storybook Controls enables you to interact with a component's inputs dynamically, without needing to code. It creates an addon panel for you to edit inputs live.

Stories written for controls are:
```

Reviewer 4: Address these objections: 
- "Do I have to change the way I write components for Controls?"
- "Do controls and args lock me into an API?"

The answers can be yes, but it helps to know straightaway.

- [Migrating from knobs](#migrating-from-knobs)
  - [Knobs to auto-generated args](#knobs-to-auto-generated-args)
  - [Knobs to manually-configured args](#knobs-to-manually-configured-args)
- [Configuration](#configuration)
  - [Control annotation](#control-annotation)
  - [Parameters](#parameters)
    - [expanded](#expanded)
- [Framework support](#framework-support)
- [FAQs](#faqs)
  - [How will this replace addon-knobs?](#how-will-this-replace-addon-knobs)

## Get started

To get started with `addon-controls`:

1. [Install the addon](#installation)
2. [Write a story that uses Args](#writing-stories)
3. [(optional) Configure the addon](#configuration)
4. [(optional) Migrate knobs stories](#migrating-from-knobs)

## Installation

Controls requires [Storybook Docs](https://github.com/storybookjs/storybook/tree/next/addons/docs). If you're not using it already, please install that first.

Next, install the package:

```sh
npm install @storybook/addon-controls -D # or yarn
```

And add it to your `.storybook/main.js` config:

```js
module.exports = {
  addons: [
    '@storybook/addon-docs'
    '@storybook/addon-controls'
  ],
};
```

## Writing stories

> Once the addon is installed you should see a `Controls` tab in the addons panel.

Reviewer 4: Tell the user what they'll accomplish in this section. 

"Let's see how to write stories that automatically generate controls... When you install the addon, you should see a Controls tab in the addons pannel."

Controls are only available for stories that make use of [Storybook Args](https://github.com/storybookjs/storybook/blob/next/docs/src/pages/formats/component-story-format/index.md#args-story-inputs), so to get interactive controls, you need to upgrade your stories.

> ### No args

Reviewer 4: If you're reading these docs for the first time, folks won't have any context for stories, much less stories with or without args. 

Instead, pretend that the reader is a brand new user who is writing stories and seeing args for the first time.

"No args" --> "Getting started"

Let's start with the following component/story combination, which should look familiar if you're coming from an older version of Storybook.

```tsx
import React from 'react';
interface ButtonProps {
  /** The main label of the button */
  label?: string;
}
export const Button = ({ label = 'FIXME' }: ButtonProps) => <button>{label}</button>;
```

> And here's a simple story that shows that Button component:

Reviewer 4: ```suggestion
And here's a story that shows that Button component:
```

```jsx
import React from 'react';
import { Button } from './Button';
export default { title: 'Button', component: Button };

export const Basic = () => <Button label="hello" />;
```

After installing the controls addon, you'll see a new tab that shows the component's props, but it doesn't show controls because the story doesn't use args. That's not very useful, but we'll fix that momentarily.

<center>
  <img src="./docs/media/addon-controls-install.png" width="100%" />
</center>

### Auto-generated args

> To upgrade your story to an Args story, simply modify it to accept an args object. **NOTE:** you may need to refresh the browser at this point.

Reviewer 4: ```suggestion
To upgrade your story to an Args story, modify it to accept an args object. **NOTE:** you may need to refresh the browser at this point.
```

```jsx
export const Basic = (args) => {
  console.log({ args });
  return <Button label="hello" />;
};
```

Now you'll see auto-generated controls in the `Controls` tab, and you can see the `args` data updating as you edit the values in the UI:

<center>
  <img src="./docs/media/addon-controls-args-logging.png" width="100%" />
</center>

> Since the args directly matches the `Button`'s props, we can simply pass it into the args directly:

Reviewer 4: ```suggestion
Since the args directly matches the `Button`'s props, we can pass it into the args directly:
```

);
Reflow.args = { count: 3, label: 'reflow' };
```

This generates the following UI:

<center>
  <img src="./docs/media/addon-controls-args-background-reflow.png" width="100%" />
</center>

Storybook has inferred the control to be a numeric input based on the initial value of the `count` arg. As we did above, we can also specify a custom control [as we did above](#custom-controls). Only this time since it's story specific we can do it at the story level:

```jsx
// export const Reflow = ... (as above)
// Reflow.args = ...
Reflow.argTypes = {
  count: { control: { type: 'range', min: 0, max: 20 } },
};
```

This generates the following UI with a custom range slider:

<center>
  <img src="./docs/media/addon-controls-args-background-reflow-slider.png" width="100%" />
</center>

### Template stories

Suppose you've created the `Basic` story from above, but now we want to create a second story with a different state, such as how the button renders with the label is really long.

The simplest thing would be to create a second story:

```jsx
export const VeryLongLabel = (args) => <Button {...args} />;
VeryLongLabel.args = { label: 'this is a very long string', background: '#ff0' };
```

> This works, but it repeats code. What we really want is to reuse the `Basic` story, but simply create a different initial state. In Storybook we do this idiomatically for Args stories:

Reviewer 4: ```suggestion
This works, but it repeats code. What we want is to reuse the `Basic` story, but with a different initial state. In Storybook we do this idiomatically for Args stories:
```

```jsx
export const VeryLongLabel = Basic.bind();
VeryLongLabel.args = { label: 'this is a very long string', background: '#ff0' };
```

We can even reuse initial args from other stories:

```jsx
export const VeryLongLabel = Basic.bind();
VeryLongLabel.args = { ...Basic.args, label: 'this is a very long string' };
```

<center>
  <img src="./docs/media/addon-controls-args-template.png" width="100%" />
</center>

> ## Migrating from knobs

Reviewer 4: I suggest we move this section to the FAQ. Something like "How to migrate from knobs". This isn't necessary for folks getting started with Controls and will quickly lose relevance as controls is widely adopted (which means we'll have to move it anyways).

|             | inline-radio | inline radio buttons input                                     |                |
|             | check        | multi-select checkbox input                                    |                |
|             | inline-check | multi-select inline checkbox input                             |                |
|             | select       | select dropdown input                                          |                |
|             | multi-select | multi-select dropdown input                                    |                |
| **string**  | text         | simple text input                                              | -              |
|             | color        | color picker input that assumes strings are color values       | -              |
|             | date         | date picker input                                              | -              |

### Parameters

Controls supports the following configuration parameters, either [globally or on a per-story basis](https://storybook.js.org/docs/basics/writing-stories/#parameters):

- [expanded](#expanded)

> #### expanded

Reviewer 4: Change the title to "Show property documentation"

This feature is amazing! TIL
